apis/v1alpha2/grpcroute_types.go

```diff
@@ -0,0 +1,461 @@
+	// MUST prioritize rules based on the following criteria, continuing on
+	// ties. Merging must not be done between GRPCRoutes and HTTPRoutes.
+	// Precedence must be given to the rule with the largest number of:
+	//
+	// * Characters in a matching non-wildcard hostname.
+	// * Characters in a matching hostname.
+	// * Characters in a matching service.
+	// * Characters in a matching method.
+	// * Header matches.
+	//
+	// If ties still exist across multiple Routes, matching precedence MUST be
+	// determined in order of the following criteria, continuing on ties:
+	//
+	// * The oldest Route based on creation timestamp.
+	// * The Route appearing first in alphabetical order by
+	//   "{namespace}/{name}".
+	//
+	// If ties still exist within the Route that has been given precedence,
+	// matching precedence MUST be granted to the first matching rule meeting
+	// the above criteria.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{method: {type: "Exact"}}}
+	Matches []GRPCRouteMatch `json:"matches,omitempty"`
+
+	// Filters define the filters that are applied to requests that match
+	// this rule.
+	//
+	// The effects of ordering of multiple behaviors are currently unspecified.
+	// This can change in the future based on feedback during the alpha stage.
+	//
+	// Conformance-levels at this level are defined based on the type of filter:
+	//
+	// - ALL core filters MUST be supported by all implementations.
+	// - Implementers are encouraged to support extended filters.
+	// - Implementation-specific custom filters have no API guarantees across
+	//   implementations.
+	//
+	// Specifying a core filter multiple times has unspecified or custom
+	// conformance.
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Filters []GRPCRouteFilter `json:"filters,omitempty"`
+
+	// BackendRefs defines the backend(s) where matching requests should be
+	// sent.
+	//
+	// An `UNIMPLEMENTED` status MUST be returned if there are no BackendRefs or filters
+	// specified that would result in a response being sent.
+	//
+	// A BackendRef is considered invalid when it refers to:
+	//
+	// * an unknown or unsupported kind of resource
+	// * a resource that does not exist
+	// * a resource in another namespace when the reference has not been
+	//   explicitly allowed by a ReferenceGrant (or equivalent concept).
+	//
+	// When a BackendRef is invalid, `UNIMPLEMENTED` statuses MUST be returned for
+	// requests that would have otherwise been routed to an invalid backend. If
+	// multiple backends are specified, and some are invalid, the proportion of
+	// requests that would otherwise have been routed to an invalid backend
+	// MUST receive an `UNIMPLEMENTED` status.
+	//
+	// When a BackendRef refers to a Service that has no ready endpoints, it is
+	// recommended to return an `UNAVAILABLE` status.
+	//
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
```

Comment:
Could you see about updating this to be similar to the new(ish) HTTPRoute BackendRefs field reference (https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/httproute_types.go#L201-L231) here?

In particular, I think that we need to be clear about the status behavior; When a BackendRef is invalid, the Attached Condition and ResolvedRefs Conditions should be updated in a similar way to the HTTPRoute ones.

In particular, I think we need to be specific about what happens if zero, one out of one, or one out of more than one BackendRefs are invalid.

Reply:
Done.